### PR TITLE
storage: filesystem/dotgit, Improve load packed-refs

### PR DESCRIPTION
**Pull request by fcharlie**
In this PR I tweaked the parsing strategy of packed-refs to not continue if it is a single reference if it has already been found, which is very useful for repositories with thousands of references.
This reduces one iteration of references for loading all references, and also reduces memory allocations for a large number of references.

### CODE REVIEW

File: plumbing/reference.go [hunk 1]
```diff
@@ -222,3 +222,11 @@ func (r *Reference) String() string {
...
 	return v.String()
 }
+
+// ReferenceSlice attaches the methods of sort.Interface to []*Reference, sorting in
+// increasing order.
+type ReferenceSlice []*Reference
+
+func (p ReferenceSlice) Len() int           { return len(p) }
+func (p ReferenceSlice) Less(i, j int) bool { return p[i].Name() < p[j].Name() }
+func (p ReferenceSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
```
Comment (pjbgf) [inline plumbing/reference.go:232]: This seems to be orthogonal to the PR's objective. Do you mind splitting this into its own PR - with some tests.

```suggestion
```

File: storage/filesystem/dotgit/dotgit.go [hunk 2]
```diff
@@ -780,33 +788,23 @@ func (d *DotGit) RemoveRef(name plumbing.ReferenceName) error {
...
-	for _, ref := range packedRefs {
+	return d.findPackedRefs(func(ref *plumbing.Reference) bool {
 		if !seen[ref.Name()] {
 			*refs = append(*refs, ref)
 			seen[ref.Name()] = true
 		}
-	}
-	return nil
+		return true
+	})
```
Comment (pjbgf) [inline storage/filesystem/dotgit/dotgit.go:797]: The implementation of `func` here seems the same as the one used on line `801`. We could extract this as a standalone func and use the same code in both places.

File: storage/filesystem/dotgit/dotgit.go [hunk 3]
```diff
@@ -780,33 +788,23 @@ func (d *DotGit) RemoveRef(name plumbing.ReferenceName) error {
...
 
 func (d *DotGit) addRefsFromPackedRefsFile(refs *[]*plumbing.Reference, f billy.File, seen map[plumbing.ReferenceName]bool) (err error) {
-	packedRefs, err := d.findPackedRefsInFile(f)
-	if err != nil {
-		return err
-	}
-
-	for _, ref := range packedRefs {
+	return d.findPackedRefsInFile(f, func(ref *plumbing.Reference) bool {
 		if !seen[ref.Name()] {
```
Comment (pjbgf) [inline storage/filesystem/dotgit/dotgit.go:802]: Although line `739` checks for `nil`, this func should also check for it. The reasoning being that future execution paths may assume it would be safe to call in with a nil `ref`.

File: storage/filesystem/dotgit/dotgit.go [hunk 4]
```diff
@@ -718,48 +718,56 @@ func (d *DotGit) Ref(name plumbing.ReferenceName) (*plumbing.Reference, error) {
...
 	s := bufio.NewScanner(f)
-	var refs []*plumbing.Reference
 	for s.Scan() {
 		ref, err := d.processLine(s.Text())
 		if err != nil {
-			return nil, err
+			return err
 		}
 
 		if ref != nil {
```
Comment (pjbgf) [inline storage/filesystem/dotgit/dotgit.go:729]: Once all implementations of `recv` checks for `nil` and short-circuit for that scenario, this `if` becomes redundant.

**Review by pjbgf — COMMENTED** [review]
Please reword the commit to align with [contributing guidelines](https://github.com/go-git/go-git/blob/master/CONTRIBUTING.md). Or refer to the new PR's title.

**Comment by fcharlie** [comment]
> Please reword the commit to align with [contributing guidelines](https://github.com/go-git/go-git/blob/master/CONTRIBUTING.md). Or refer to the new PR's title.
> 
> Please add a test to safeguard the changes against regression. It would be also nice if you could paste the benchmark results in the PR to highlights the memory allocation reductions the changes will achieve.

I don't have enough time to perfect the test for the time being, here is a simple case:

```pwsh
# windows
 git clone --bare https://github.com/git/git.git "$env:TEMP/git.git"
```

```go
func TestFindRef(t *testing.T) {
	now := time.Now()
	repo, err := git.PlainOpen("path/to/git")
	if err != nil {
		fmt.Fprintf(os.Stderr, "open error: %v\n", err)
		return
	}
	ref, err := repo.Reference("refs/heads/master", false)
	if err != nil {
		fmt.Fprintf(os.Stderr, "open error: %v\n", err)
		return
	}
	fmt.Fprintf(os.Stderr, "%s %s use time: %v\n", ref.Hash(), ref.Name(), time. Since(now))
}
```

refs:
```bash
# git for windows
git show-ref|wc -l
907
```

before:

image

>7580f92ffa970b9484ac214f7b53cec5e26ca4bc refs/heads/master use time: 1.6423ms

after:

image

>7580f92ffa970b9484ac214f7b53cec5e26ca4bc refs/heads/master use time: 529.5µs

**Comment by pjbgf** [comment]
The benchmarks shows that the changes results in a more predictable time cost per op:

```
goos: linux
goarch: amd64
pkg: github.com/go-git/go-git/v5/storage/transactional
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz

/tmp/current
BenchmarkReference-16                 1        1807192975 ns/op        151665280 B/op    520597 allocs/op
BenchmarkReference-16                 1        2822611253 ns/op        148277904 B/op    482159 allocs/op
BenchmarkReference-16                 1        1871437889 ns/op        148332872 B/op    481993 allocs/op
BenchmarkReference-16                 1        2013824760 ns/op        147963112 B/op    482186 allocs/op
BenchmarkReference-16                 1        3188275241 ns/op        148189280 B/op    482150 allocs/op
BenchmarkReference-16                 1        1929030106 ns/op        148297832 B/op    481988 allocs/op

/tmp/after
BenchmarkReference-16                 1        1298677561 ns/op        150877264 B/op    520869 allocs/op
BenchmarkReference-16                 1        1170506662 ns/op        147997344 B/op    482935 allocs/op
BenchmarkReference-16                 1        1105562172 ns/op        147640968 B/op    482887 allocs/op
BenchmarkReference-16                 1        1056970154 ns/op        148353736 B/op    482919 allocs/op
BenchmarkReference-16                 1        1565437569 ns/op        148816872 B/op    482753 allocs/op
BenchmarkReference-16                 1        1095235494 ns/op        148314880 B/op    482941 allocs/op
```

benchstat
```
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
              │ /tmp/current │             /tmp/after             │
              │    sec/op    │   sec/op     vs base               │
BenchmarkReference-16    1.971 ± 62%   1.138 ± 38%  -42.27% (p=0.002 n=6)

              │ /tmp/current │          /tmp/after           │
              │     B/op     │     B/op      vs base         │
BenchmarkReference-16   141.4Mi ± 2%   141.5Mi ± 2%  ~ (p=0.818 n=6)

              │ /tmp/current │            /tmp/after             │
              │  allocs/op   │  allocs/op   vs base              │
BenchmarkReference-16    482.2k ± 8%   482.9k ± 8%  +0.16% (p=0.041 n=6)
```

Code used:
```go
func BenchmarkReference(b *testing.B) {
	repo, err := git.PlainClone(b.TempDir(), true, &git.CloneOptions{
		URL: "https://github.com/go-git/go-git.git",
	})
	if err != nil {
		b.Fatalf("open error: %v", err)
	}

	for i := 0; i < b.N; i++ {
		_, err := repo.Reference("refs/heads/master", false)
		if err != nil {
			b.Fatalf("open error: %v", err)
		}
	}
}
```

**Review by pjbgf — APPROVED** [review]
@fcharlie thanks for working on this. :bow:
